preload.sh

```diff
@@ -13,76 +13,188 @@ API_HOST=${API_HOST:-"https://api.resin.io"}
+  local NEW_PART_END=$(expr $PART_END + $2)
+  echo "Resize partition $1 from offset ${PART_END}MB to ${NEW_PART_END}MB"
+  # Resize partition table
+  # Both extended and logical partition must be increased
+  parted -s "$IMAGE" resizepart 4 "${NEW_PART_END}MB" resizepart 6 "${NEW_PART_END}MB"
+}
 
-# Calculate new partition end by getting current partition end and adding the additional spzce.
-PART_END=$(parted -s -m "$IMAGE" p | tail -n 1 | awk -F ':' '{print $3 + '$IMG_ADD_SPACE'}')
+# Expand the image
+# NOTE: Depends on get_app_data() being called first
+function expand_image() {
+  CONTAINER_SIZE=$(curl -s "$REGISTRY_HOST/v1/images/$IMAGE_ID/ancestry" | \
+  jq '.[]' | awk '{print "'$REGISTRY_HOST'/v1/images/" $1 "/json"}' | \
+  xargs -r -n 1 curl -I -s | \
+  grep 'X-Docker-Size' | \
+  awk '{s+=$2} END {print int(s / 1000000)}')
+  echo "Container size:" $CONTAINER_SIZE "MB"
+  # Size will be increased by 110% of container size
+  IMG_ADD_SPACE=$(expr $CONTAINER_SIZE / 100 + 300)
+  echo "Expanding image by" $IMG_ADD_SPACE "MB"
+  # Add zero bytes to image to be able to resize partitions
+  dd if=/dev/zero bs=1M count="$IMG_ADD_SPACE" >> "$IMAGE"
+}
 
-# Resize partition table
-# Both extended and logical partition must be increased
-parted -s "$IMAGE" resizepart 4 "${PART_END}MB" resizepart 6 "${PART_END}MB"
+# Get the resinOS version by mounting the root partition
+# and read `/etc/os-release` to set the variables in it
+# Example:
+# 
+# ID="resin-os"
+# NAME="Resin OS"
+# VERSION="2.0.0-beta.7"
+# VERSION_ID="2.0-beta.7"
+# PRETTY_NAME="Resin OS 2.0.0-beta.7"
+# RESIN_BOARD_REV=82eeb8b
+# META_RESIN_REV=0870520
+# SLUG=raspberrypi3
+# MACHINE=raspberrypi3
+function get_resin_os_version() {
+  local ROOT_PART_START=$(get_partition_start 2 B)
+  local ROOTMNT="/mnt/${APP_ID}-root"
+  local OS_RELEASE="${ROOTMNT}/etc/os-release"
+  echo "rootfs partition start: $ROOT_PART_START"
+  echo "mounting rootfs to: $ROOTMNT"
+  mkdir -p $ROOTMNT
+  # Make sure directory exists before mounting to it
+  sync
+  sleep 2
+  echo "mount -o loop,ro,offset=${ROOT_PART_START} $IMAGE $ROOTMNT"
+  mount -o "loop,ro,offset=${ROOT_PART_START}" "$IMAGE" "$ROOTMNT"
+  local DEVICE=$(losetup --partscan --find --show $IMAGE)
+  echo "blockdevice" $DEVICE
+  eval $(cat $OS_RELEASE)
+  # Clean up
+  test "`mount | grep \"$ROOTMNT\"`" && umount $ROOTMNT || true
+  test -d $ROOTMNT && rmdir $ROOTMNT
+  echo "Detaching $DEVICE"
+  losetup -D
+}
 
-# mount partition
+# Get app data and write to temporary file
+function get_app_data() {
+  ( if test "$API_TOKEN"; then
+          curl -sH "Authorization: Bearer $API_TOKEN" "$API_HOST/ewa/application($APP_ID)?\$expand=environment_variable"
+  elif test "$API_KEY"; then
+          curl "$API_HOST/ewa/application($APP_ID)?\$expand=environment_variable&apikey=$API_KEY"
+  fi ) | jq --arg registryHost "$REGISTRY_HOST" '.d[0] |
+          (.git_repository | split("/") | .[1] | rtrimstr(".git")) as $repoName |
```

Comment:
If you switch to `$API_HOST/v2/` instead of `$API_HOST/ewa/` you can drop the `rtrimstr(".git")`